### ISSUE TITLE
Unified "Read" Button Dropper

### ISSUE THREAD

**Issue by cdrini** [internetarchive/openlibrary#5831]
We have a lot of read-y actions a user can perform; designing these as a dropdown would allow us to expose these options in more places; eg search results, carousels; without taking up a bunch of space.

## Options

Borrow / Read, Listen, Search Inside, Download, Locate (meaning check nearby libraries) 

## Description, Patron Story
- What is the design change?

Turns the read button into a drop-down menu with more options

- Why is this design change necessary?

saves real estate on the page by grouping similar options (downloads, etc)

- Who is it for?

all patrons and stakeholders of the books page

- constraints and implementation

Preferably everything is managed using `details` html element and no js is required, all css.

## Relevant Designs

![image](https://github.com/internetarchive/openlibrary/assets/6251786/06f0597d-f331-46b5-8e1a-7bd2ef2bad0e)

<img width="383" alt="Screenshot 2024-05-10 at 8 33 10 AM" src="https://github.com/internetarchive/openlibrary/assets/978325/e433839e-3f9b-471d-9e2e-60634a3b6bf2">

https://www.figma.com/file/dYQkIJOJeMo9hx7ewZQH9x/Open-Library%3A-Design-System-%26-Component-Library?type=design&node-id=601-199&mode=design&t=giRw2crBrKKYkWGi-0

### Problem Statement

* Each book has one or more different applicable actions for reading, locating a copy, etc.
  * e.g. Read, Borrow, Listen, Download epub, Search Inside, Check local library, Purchase...
* In some case, we overload the read button with an :headphones: icon to suggest there's also a listen option. This is a confusing pattern. 
* Trying to represent all a book's options risks taking up space, introducing multiple access patterns, and creating multiple locations where a patron may have to learn to look.

### Proposal & Constraints
Actions in dropdown (WIP):

* Borrow (or) Read (or) Listen (or) Borrow & Listen
* Download epub
* Search Inside
* Purchase
* Check local library

### Stakeholders
@mekarpeles  @cdrini

**Comment by seabelis** [comment]
Related request from a patron:
"Hi! Please add "Want to read" button to search resulta view so that results don't have to opened separately."

**Comment by jimchamp** [comment]
Decisions made be staff about this:
1. We want to avoid ever having to show "Not in Library" as a call to action for this button.
2. We want to include the library links in the redesigned "Read" button.  Presently, library links are found in the second item in the sidebar (highlighted below):

![library_links](https://user-images.githubusercontent.com/28732543/166836135-46c6a0a7-eaef-4b64-90e8-3f1ff72b91b0.png)

**Comment by mekarpeles** [comment]
See: https://www.figma.com/file/nVmPYLTaxWWYchVe2ArVbl/Unified-Borrow-Button?node-id=0%3A1

**Comment by mekarpeles** [comment]
## Description, Patron Story
- What is the design change?

Turns the read button into a drop-down menu with more options

- Why is this design change necessary?

saves real estate on the page by grouping similar options (downloads, etc)

- Who is it for?

all patrons and stakeholders of the books page

## Relevant Designs


<img width="466" alt="Screenshot 2024-01-11 at 9 00 26 AM" src="https://github.com/internetarchive/openlibrary/assets/978325/f6ae1312-0ae1-4857-b363-9cc37c874f37">


## Feedback Questions

- What options are missing from the dropdown? e.g. trusted book providers? vendor/prices?

- how do we feel about the choice to expand/collapse the 

- how should search inside work, e.g. should it also expand and show a search box one can type in?

- concerns? are we hiding things and making them less discoverable or usable?

## Decisions


### Stakeholders

**Comment by cdrini** [comment]
Here's an iteration on the designs in Mek's figma. We talked about this during our design call and are still uncertain on whether we want it, but if we did, here are some more designs!

![image](https://github.com/internetarchive/openlibrary/assets/6251786/06f0597d-f331-46b5-8e1a-7bd2ef2bad0e)

https://www.figma.com/file/dYQkIJOJeMo9hx7ewZQH9x/Open-Library%3A-Design-System-%26-Component-Library?type=design&node-id=1382-2&mode=design&t=IWiihHe1SPc4LcSm-4

**Comment by mekarpeles** [comment]
**Resolved**: We're going to move forward with an implementation based on html `details` to have a Read button dropdown which at minimum includes:
* Read / borrow
* Listen
  * we're going to remove the slide-out headphones interaction on borrow button
  * Listen 7 days: 3.5k + 650 = ~4.2k https://plausible.prod.archive.org/openlibrary.org?goal=CTAClick
* Search inside
  * for now, we're going to remove the search inside button from the book page and see how analytics change
* Download options
* Locate --> worldcat
  * we should also keep download options and locate sections on the book page for now to see which get clicked -- needs additional analytics.

**Comment by mekarpeles** [comment]
Here's a codepen with scaffolding for such a redesign: https://codepen.io/mekarpeles/pen/XWQyjYX

**Comment by mekarpeles** [comment]
@SivanC how do you feel about tag teaming on this one?

**Comment by SivanC** [comment]
Great, I'll start this week and keep you posted on my progress!

**Comment by SivanC** [comment]
@mekarpeles Hi, I've made [a codepen of my own](https://codepen.io/sivanc/pen/VwOpXwq?editors=1100) that is a little closer to the Figma design, but there's two big issues I'm still trying to solve:
- If the dropper content is wide enough, the dropper button will become wider as well, not sure how to make it so that the content extends past the button, like in the mockup
- I can't seem to make the bottom left and right corners of the main button and dropper rounded, respectively

While researching, I also noticed that there is much discussion around the accessibility of nested interactive elements in `summary` elements, which I can share if you're interested. Supposedly this style of "details menu" has been popularized by Github but I haven't found much to back that up.

Please let me know what you think! I'll continue to tweak the codepen (especially to try and implement the light blue sub-buttons for e.g. Download) in the coming week and see if I can find answers to the above issues.

**Comment by mekarpeles** [comment]
@SivanC this is looking fantastic :) really nice work

You could make the `cta-content-inner` `position: absolute` which will make its width independent from its parent

It looks re: border radius, you've been able to get the bottom left and right corners rounded correctly :)

**Comment by SivanC** [comment]
Thanks for the tip, that actually fixes both of the issues! Without `position: absolute` the corners of the main/dropper button are still unrounded when the dropper is expanded.

**Comment by mekarpeles** [comment]
The LoanStatus macro handles all of the rending of primary action buttons on the website:
https://github.com/internetarchive/openlibrary/blob/master/openlibrary/macros/LoanStatus.html#L43

It makes use of the `ReadButton` macro, which is the one we want to upgrade:
https://github.com/internetarchive/openlibrary/blob/master/openlibrary/macros/ReadButton.html

**Comment by SivanC** [comment]
@mekarpeles @jimchamp Hi, sorry for my relative radio silence recently, just wanted to give you guys an update on my progress in migrating my codepen design to my local environment.

Here's what it looks like:
Closed:
![image](https://github.com/internetarchive/openlibrary/assets/17033120/9f670997-065f-4f31-bd2f-37bb0e99dacb)
Open:
![image](https://github.com/internetarchive/openlibrary/assets/17033120/c5ed7910-a5eb-437b-93ab-a00b167fdec2)

Styling issues I'm currently working on:
1. I tried to read up on specificity and how to override certain style rules in the best way but I'm a little stumped. The two fixes I need to make are to remove the -5px horizontal margin from the `summary` element (causing the current misalignment) and to set `list-style(-type): none` for the dropdown content to remove the bullets.

2. The worst of the issues is the weird layout of the read button with the `summary::after` pseudo-element. Not sure why it's so wonky when it worked fine in the codepen. I tried putting a variety of display styles on a variety of elements to no avail. You can see that my element inspector shows a strange (to me) occupied space with `display: flex` enabled for the `summary` element:
With `display: flex`:
![image](https://github.com/internetarchive/openlibrary/assets/17033120/a46b3000-520c-4908-b34e-4d53608ca147)
Without flex:
![image](https://github.com/internetarchive/openlibrary/assets/17033120/187f4a4f-0135-4827-986e-f687b9983fd1)

Other issues I can probably solve by myself but are still WIP are coloring the dropdown arrow and adjusting spacing and hover behavior on the list elements in the dropdown. I'll also begin to work on icons, button functionality and the like when these issues are out of the way. I'll make sure to summarize this again on the call tomorrow. Thanks for taking a look!

**Comment by rebecca-shoptaw** [comment]
@SivanC Dropping by with some CSS help in case it's still needed!
1. For list-style troubleshooting, you'll want to make sure you're adding `list-style-type: none` to whichever element is the `ul`
2. I'd recommend adding a `flex-grow: 1` to the "Read" part of the button, which makes it automatically take up all the remaining space in the flex container -- if this doesn't work, you can try pairing it with a `flex-shrink: 0` and `flex-basis: 1` as well! 🙂

**Comment by SivanC** [comment]
@rebecca-shoptaw Hi thanks for your tips! As far as the list styling the issue is that I can see in my element inspector that my styling is being overwritten by other rules; other than that it's being applied correctly.
![image](https://github.com/internetarchive/openlibrary/assets/17033120/f67ed3cc-51de-4351-a4f1-160d27ff8d9d)

**Comment by rebecca-shoptaw** [comment]
@SivanC Aha! You want to make sure you're applying it to the full list `ul` element rather than each `li` element -- i.e. if `.cta-dropper__content-list` is the class for your `ul` you want to give the `list-style-type: none` to that. 

If that doesn't work, there are some specificity tricks you can try, but fingers crossed that does the trick!

**Comment by SivanC** [comment]
@rebecca-shoptaw Unfortunately the reason I had it on the `li` elements was because I thought the same initially and applied it to the `ul` to no effect. `cta-dropper__content-list` is the class associated with the `ul` element and while the styling shows up in the inspector the bullet points remain:
![image](https://github.com/internetarchive/openlibrary/assets/17033120/c6f1385f-0d0e-4546-8921-b2eb19281214)
![image](https://github.com/internetarchive/openlibrary/assets/17033120/6020de00-9605-4a9e-bd4c-bbd371c8dec8)

Here's the full HTML right now:
```
<div class="dropdown-button-group">
  <div class="cta-dropper">
    <details class="cta-dropper__details">
      <summary class="cta-dropper__summary">
        <div class="cta-button-group">
          <a href="$(stream_url)" title="$title" class="cta-btn cta-btn--ia cta-btn--available cta-btn--$(action)"
              $:analytics_attr(analytics_action)
              $if loan:
                data-userid="$(loan['userid'])"
              >$label</a>
        </div>
      </summary>
      <div class="cta-dropper__content">
        <ul class="cta-dropper__content-list">
          <li>Listen</li>
          <li>Search Inside</li>
          <li>Check Nearby Libraries</li>
        </ul>
      </div>
    </details>
  </div>
</div>
```

**Comment by rebecca-shoptaw** [comment]
@SivanC Thank you!! Ok, it looks like the styling is being inherited because the parent element of all of this has a `read-options` class, which has its own `li` rules.

This means you'll want to use selector specificity to overrule the existing `li` rules in this particular case. It looks like someone else has already encountered a similar issue and fixed it in `read-panel.less`:
https://github.com/internetarchive/openlibrary/blob/9fc1f69e86bbb0ef3fb3595d27b23b0a59d691d2/static/css/components/read-panel.less#L164-L167

So it seems that the simplest course of action would be to add a few lines like this directly above those:
```
.read-options .cta-dropper__content-list li {
  list-style: none;
}
```
Effectively you're saying that while all `.read-options` `li` elements usually have a `list-style: disc`, if they are the child of a `.cta-dropper__content-list` they have a `list-style: none`.

There may be a more elegant way to do this, but this should do the trick!

**Comment by SivanC** [comment]
@rebecca-shoptaw I was about to respond that nothing I tried was working ... until! I noticed that both `.read-options` and `.Tools .btn-notice` asked for disc style, so I made a rule that covered both of them according to your suggestion:
```
.read-options .cta-dropper__content-list li, .Tools .btn-notice .cta-dropper__content-list li {
  list-style: none;
}
```
![image](https://github.com/internetarchive/openlibrary/assets/17033120/a04fd053-8357-4a60-aa4a-34a3f665cff9)

No more bullet points! :D

**Comment by SivanC** [comment]
Update as of 7/26/24 design call, clarifying some questions surrounding dropper content:
- Search inside will be a input field
- Clicking “Download” will scroll to / highlight the Download options (ala clicking on ratings under author name)
- Clicking Download pills will download directly
- Buy this book should not be in the dropper for now; considering out of scope
- Should still show “not in library” + grey for now in this PR
- In the editions box on a work’s page, read buttons should still have droppers, clicking e.g. download should take you to the download section for that edition (scroll + pulsing border)

**Comment by SivanC** [comment]
Long overdue update. Since the last comment, the focus has been on getting the overall dropper into a good state visually, and getting the Listen, Search Inside, and Locate buttons working. Download will come after this version has been launched.

"Check Nearby Libraries" button has been turned into Locate button, and made into its own (merged) PR (#9953). This is now live, and always directs the user to Worldcat, now using a more robust system to more reliably find Worldcat's entry for the edition using ISBN/OCLC and then title as a fallback.

Next steps:
* [ ] Make sure dropper button is fully functional for print-disabled users (includes making a new Search Inside main CTA, styling will be fairly complex in my estimation)
* [ ] Integrate Locate button into dropper. This might have some effects on the existing Locate button spacing. Overall probably not too difficult since it was designed as part of the dropper.
* [ ] At this point I think the dropper will be ready functionally, but will probably need some visual touchups, which hopefully won't take very long either.
* [ ] Testing and any final changes that need to be made

After the initial launch, the broad next steps are:
* [ ] Add download button
* [ ] Add dropper to carousels (might require big changes to carousels)

**Comment by mekarpeles** [comment]
@SivanC, just for v1 (to go quicker), can we keep "Search Inside" a separate button?

On books page, we'd see the Unified Read Button and then the Search Inside below it (as is)

On search results, if preview only, just show search inside. Previews will be handled separately by #10199

<img width="252" alt="Screenshot 2024-12-29 at 2 58 22 PM" src="https://github.com/user-attachments/assets/6fe930e0-e8f7-41a0-be49-1b7724586971" />

**Comment by SivanC** [comment]
Update of remaining issues before testing as of 01/16/25, in rough order of importance:
- In mobile view, dropper menu is not visible in edition menu (blocked by `overflow: auto` in `static/css/base/common.less` line 108), and styling of the dropper menu is broken
- Clicking the dropdown arrow on the search page also selects the book as if you clicked elsewhere on the search item. I believe the issue is caused because the search item is an ancestor of the dropdown arrow, and both have pointer events enabled. Maybe can be fixed by adding a div behind the arrow
- In tablet and mobile view, the dimensions of the dropper's cta button are incorrect in the edition table. Can probably be fixed by adjusting the placement of `cta-button-group` in `macros/ReadButton.html`, but doing that will require rewriting some other CSS
- Colored area of dropdown buttons on hover does not exactly match button area (i.e. there are some areas of each button that are not properly darkened on hover, not very noticeable).
- (less important) A legacy CSS rule that was inhibiting styling of the dropper menu uses `#searchResults ul`, which cannot be overridden without violating specificity rules. Doesn't affect visuals/UX in any way
- (less important) With the removal of `.Tools` from the `.Tools .read-options` list of CSS rules, the preview button's styles were being overwritten. This has been fixed by making them more specific (as they were originally, just at a deeper nested level), which is/was too specific according to the linter. Maybe needs to be reworked but doesn't affect visuals/UX in any way
